### ISSUE TITLE
Add runtime module to enable concurrent load of manifest files.

### ISSUE THREAD

**Issue by liurenjie1024** [apache/iceberg-rust#124]
Currently we implement manifest loading in a sequential approach, e.g. load them one by one. We should add load them concurrently. This requires submitting tasks to rust async runtime, and we should be careful to runtime agnostic.

**Comment by odysa** [comment]
Hi, is this what you refer to? Can you plz explain more about "careful to runtime agnostic"? Is there anything we need to be careful when implementing concurrent scanning?
https://github.com/apache/iceberg-rust/blob/9768b0ef1c7ab3631f69ee52d18ea928e4a4250f/crates/iceberg/src/scan.rs#L140-L145

**Comment by liurenjie1024** [comment]
> Hi, is this what you refer to?

Yes, exactly.

> Can you plz explain more about "careful to runtime agnostic"? Is there anything we need to be careful when implementing concurrent scanning?

I mean we may need an extra layer for task scheduling, so that we can be adopted to any async runtime such as tokio, async-std.

**Comment by odysa** [comment]
> I mean we may need an extra layer for task scheduling, so that we can be adopted to any async runtime such as tokio, async-std.

Do you want users to choose their own runtime like [sqlx](https://github.com/launchbadge/sqlx/tree/main#install)?
They are building an abstraction layer([Runtime](https://github.com/launchbadge/sqlx/tree/next/sqlx-core/src/runtime)) so `sqlx` can run on many blocking/non-blocking runtime.
```
# tokio (no TLS)
sqlx = { version = "0.7", features = [ "runtime-tokio" ] }
# async-std (no TLS)
sqlx = { version = "0.7", features = [ "runtime-async-std" ] }
```
I am interested in this feature, but it will take some time for me to draft a design.

**Comment by odysa** [comment]
> Do you want users to choose their own runtime like [sqlx](https://github.com/launchbadge/sqlx/tree/main?rgh-link-date=2024-02-02T17%3A02%3A32Z#install)? They are building an abstraction layer([Runtime](https://github.com/launchbadge/sqlx/tree/next/sqlx-core/src/runtime?rgh-link-date=2024-02-02T17%3A02%3A32Z)) so `sqlx` can run on many blocking/non-blocking runtime.

Follow up on this. The `sqlx` uses a relatively simple [solution](https://github.com/launchbadge/sqlx/tree/main/sqlx-core/src/rt). For example, the `spawn` function.

https://github.com/launchbadge/sqlx/blob/84d576004c93a32133688426eacb50434bb5c5f0/sqlx-core/src/rt/mod.rs#L66-L74

**Comment by liurenjie1024** [comment]
> Do you want users to choose their own runtime like [sqlx](https://github.com/launchbadge/sqlx/tree/main#install)?

Yes, exactly. I don't think we should bind to some specific runtime.

> Follow up on this. The sqlx uses a relatively simple [solution](https://github.com/launchbadge/sqlx/tree/main/sqlx-core/src/rt). 

I agree that we may need to think about it carefully, sqlx's solution can only use current runtime.

> I am interested in this feature, but it will take some time for me to draft a design.

Yeah, welcome to contribute, we can work together on this.

**Comment by marvinlanhenke** [comment]
@odysa 
Just to follow up on this, any progress regarding some design ideas?

@liurenjie1024 
Do we have any reference implementation where we can get "inspired" on how things could be done, 
or do you have a particular design idea already in mind?

Is this something we should track in #348 perhaps for the v.0.4.0 release?

**Comment by marvinlanhenke** [comment]
in order to verify my understanding and possibly kick of a design discussion, we could follow the approach of `sqlx`:

- have a `runtime.rs` 
  - to define a `Runtime` trait
  - based on a feature-flag import / re-export the implementors of that trait
- have a /runtime module with specific runtime implementations
  - that implement the `Runtime` trait

For our particular use-case (loading manifest / or DataFiles on multiple threads) we could e.g. wrap `tokio::spawn` with our runtime and we're good to go? Then the `Runtime` trait would evolve to support more "use-cases"?

**Comment by marvinlanhenke** [comment]
... so as a first step - simple wrap tokio::spawn (for example) like [here](https://github.com/launchbadge/sqlx/blob/main/sqlx-core/src/rt/mod.rs#L61-L78) - and not even use a feature-flag for now; just the most simplest layer of abstraction?

@liurenjie1024
Have you already made up your mind; about how we want to read & filter the manifest files concurrently? The ideas I want to discuss would be (if they makes sense at all):

Design 1:
- iterate over manifest_list.entries() in an async stream on a single thread
- apply manifest_evaluator
- define `fn handle_data_file` where expression & inclusive_metrics evaluator are applied and return a stream of FileScanTask
- spawn `fn handle_data_file` on multiple threads

Design 2:
- basically the same as 1
- but also spawn tasks for handling manifest_list.entries()?

**Comment by Fokko** [comment]
With Iceberg, the manifests are written to a target size (8 megabyte) by default. Each manifest is bound to the same schema and partition, so you can re-use the evaluators here. I would not go overboard with the parallelism, and just create one task per manifest, and not spawn a task per manifest-entry.

**Comment by marvinlanhenke** [comment]
> How do you feel starting with one task for one manifest file

you mean: 
- spawn a new task for each manifest, load the manifest (entry.load_manifest(...).await?)
- and handle DataFiles synchronously on the same task

so if we have a manifest_list with e.g. 5 entries, 1 is pruned (ManifestEvaluator) we'd effectively spawn 4 tasks, to load the manifest and handle all the data files; is this correct?

**Comment by Fokko** [comment]
> so if we have a manifest_list with e.g. 5 entries, 1 is pruned (ManifestEvaluator) we'd effectively spawn 4 tasks, to load the manifest and handle all the data files; is this correct?

That is correct 👍 I think there might be some confusion around the naming. In the spec we have the Manifest List that contains Manifests. Within the Manifest there are manifest-entries that each point to one Datafile.

**Comment by sdd** [comment]
Using `try_for_each_concurrent` here rather than just spawning in a for loop will allow us to tune the concurrncy as it accepts a max concurrent tasks argument. I'd advocate for a data-driven approach to determine a sensible default value for this, alongside the capability to override the default with optional config, probably a `with_max_concurrency()` method on the builder and some guidance in the docs.
